app/src/main/kotlin/com/wire/android/ui/home/conversations/DownloadedAssetDialog.kt

```diff
@@ -0,0 +1,100 @@
+import com.wire.android.util.getMimeType
+import com.wire.android.util.permission.WriteToExternalStorageRequestFlow
+import com.wire.android.util.saveFileDataToDownloadsFolder
+import java.io.File
+
+@Composable
+fun DownloadedAssetDialog(conversationViewModel: ConversationViewModel) {
+    val dialogState = conversationViewModel.conversationViewState.downloadedAssetDialogState
+    val context = LocalContext.current
+
+    if (dialogState is DownloadedAssetDialogVisibilityState.Displayed) {
+        val assetData = dialogState.assetData
+        val onDownloadAsset = {
+            saveFileToDownloadsFolder(dialogState.assetName, assetData, context)
+            conversationViewModel.onFileSaved(dialogState.assetName)
+            conversationViewModel.hideOnAssetDownloadedDialog()
+        }
+        val requestWriteToExternalStorageRequestFlow =
+            WriteToExternalStorageRequestFlow(context, onDownloadAsset, writeToExternalStoragePermissionLauncher =
+            rememberLauncherForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted ->
+                if (isGranted) {
+                    onDownloadAsset()
+                } else {
+                    // TODO: Implement denied permission rationale
+                }
+            })
+
+        WireDialog(
+            title = dialogState.assetName ?: stringResource(R.string.asset_download_dialog_default_title),
+            text = stringResource(R.string.asset_download_dialog_text),
+            buttonsHorizontalAlignment = false,
+            onDismiss = { conversationViewModel.hideOnAssetDownloadedDialog() },
+            optionButton2Properties = WireDialogButtonProperties(
+                text = stringResource(R.string.asset_download_dialog_open_text),
+                type = WireDialogButtonType.Primary,
+                onClick = {
+                    openAssetFile(dialogState.assetName, assetData, context) { conversationViewModel.onOpenFileError() }
+                    conversationViewModel.hideOnAssetDownloadedDialog()
+                }
+            ),
+            optionButton1Properties = WireDialogButtonProperties(
+                text = stringResource(R.string.asset_download_dialog_save_text),
+                type = WireDialogButtonType.Primary,
+                onClick = { requestWriteToExternalStorageRequestFlow.launch() }
+            ),
+            dismissButtonProperties = WireDialogButtonProperties(
+                text = stringResource(R.string.label_cancel),
+                onClick = { conversationViewModel.hideOnAssetDownloadedDialog() }
+            ),
+        )
+    }
+}
+
+fun saveFileToDownloadsFolder(assetName: String?, assetData: ByteArray, context: Context) {
+    val file = File(context.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS), "${System.currentTimeMillis()}_$assetName")
+    file.setWritable(true)
+    file.writeBytes(assetData)
+    context.saveFileDataToDownloadsFolder(file, assetData.size)
+}
+
```

Comment:
not a expert on File() but aren't we suppose to close the file after we are done so that we can garbage collect it ? 
also this is going to be executed on the MainThread